### PR TITLE
fix(worker): guard service-worker init against stale identity binding

### PR DESCRIPTION
**Pull request by pietro909**
## Summary

Closes the service-worker identity-init race where the `MessageBus` could end up bound to a different identity than the page last asked for, causing `getBoardingAddress()` / `getAddress()` to return a previous identity's address — a potential loss-of-funds path. 

### FIFO-serialize bus initialization
- `INITIALIZE_MESSAGE_BUS` now runs on a FIFO `initChain`, so concurrent inits can no longer interleave or finish out of order and bind the bus to a stale identity.
- A `pendingInitCount` gate makes the bus refuse ordinary wallet messages while any init is queued or running, so the previous wallet stops being served the moment a new init is accepted.
- Failed inits deliver an error response (page fails fast) without poisoning the next queued init; partially-started handlers are rolled back on failure.
- New `MessageBusInitializingError`  distinguishes "an init is in flight" from "never initialized". Legacy substring/`instanceof` detection stays backward-compatible.

### page-side identity boundary assertion
- `create()` and `reinitialize()` verify the worker reports the expected x-only pubkey via `GET_STATUS` before returning a usable wallet, refusing to hand back a wallet that would serve another identity's addresses.
- Compares the stable baseline identity key

## Summary by CodeRabbit

* **New Features**
  * Added explicit “message bus initializing” reporting, including new exported marker/error for init-in-flight vs not-initialized.
  * Added an identity boundary check so service-worker operations verify the worker matches the expected wallet public identity.

* **Bug Fixes**
  * Improved message-bus initialization by serializing concurrent init requests and preventing redundant re-initialization.
  * Updated wallet request retry to wait for initialization to settle, with safer handling of partial handler startup.

* **Tests**
  * Expanded service-worker and message-bus coverage for identity matching, init concurrency, and error/backoff behavior.

### CODE REVIEW

**Review by ghost — COMMENTED** [review]
## Code Review — arkade-os/ts-sdk#571

**Reviewer: Arkana (automated, aggressive)**
**Verdict: APPROVE** ✅

This is a well-designed, defense-in-depth fix for a real loss-of-funds race condition. The FIFO-serialized init chain in the worker and the page-side identity boundary assertion are both sound and complementary. The PR is production-ready.

---

### Security analysis (loss-of-funds path)

The core threat — `getBoardingAddress()` / `getAddress()` returning a **different identity's address** after a stale init rebinds the worker — is correctly mitigated at two layers:

1. **Worker side (`messageBus.ts`):** `initChain` serializes inits FIFO. `pendingInitCount` gates ordinary messages during init, preventing the old wallet from being served while a new identity is binding. This is correct.

2. **Page side (`wallet.ts`):** `assertWorkerIdentityMatches()` compares `identity.xOnlyPublicKey()` against the worker's `GET_STATUS` response after every `create()` and `reinitialize()`. Uses `sendMessageDirect` (no ping/retry/reinit) so it cannot recurse. This is the right design — it catches any binding mismatch before a usable wallet is returned.

The comparison uses the **baseline identity key** (`identity.xOnlyPublicKey()`), not the rotated receive/boarding key, so HD wallets stay correct after rotation. ✅

### Findings (all minor — none blocking)

**1. Duplicated retry loop logic — `sendMessageWithRetry` / `sendMessageWithEvents`** (`wallet.ts:833-871`)

The `isMessageBusInitializingError` backoff-and-retry block is copy-pasted identically between `sendMessageWithRetry` and `sendMessageWithEvents`. This is a maintenance hazard — consider extracting a shared helper. Not a blocker, but worth a follow-up.

**2. `attempt--` counter manipulation** (`wallet.ts:848, 880`)

When an `isMessageBusInitializingError` occurs, `attempt` is decremented so the init-wait doesn't consume the reinit budget. This is correct behavior but the `attempt--` inside a `for` loop that also does `attempt++` on the next iteration is subtle. A comment like `// net zero: this wait should not count as a reinit attempt` would help future readers. (Minor readability nit.)

**3. Bounded backoff caps at ~9.1s total wait** (`wallet.ts:175-178`)

`MAX_INIT_WAITS = 8` with `100 * 2^attempt` capped at 2s gives: 100 + 200 + 400 + 800 + 1600 + 2000 + 2000 + 2000 = ~9.1s total. This is reasonable. A pathologically slow init could still exceed this, but the caller gets a clear error. ✅

**4. `doInit` rollback is correct** (`messageBus.ts:367-377`)

Partially-started handlers are rolled back with `h.stop().catch(() => {})` and `this.initialized = false`. The `catch` suppresses stop errors (correct — we don't want rollback failures to mask the original error). The chain heals via `.catch(() => undefined)`. ✅

**5. Error superset design is backward-compatible** (`errors.ts:8-12`)

`MESSAGE_BUS_INITIALIZING = "MessageBus not initialized: initializing"` — contains `MESSAGE_BUS_NOT_INITIALIZED` as a substring, so the existing `isMessageBusNotInitializedError` detector returns `true` for it. `MessageBusInitializingError extends MessageBusNotInitializedError` so `instanceof` also works. Older callers that don't know about the new error class will treat it as retryable (re-init), which is safe — slightly wasteful but not incorrect. ✅

**6. `MessageBusNotInitializedError` constructor signature change** (`errors.ts:7`)

Changed from no-arg to `constructor(message: string = MESSAGE_BUS_NOT_INITIALIZED)`. Default parameter preserves backward compatibility for all existing `new MessageBusNotInitializedError()` call sites. ✅

**7. `assertWorkerIdentityMatches` — `workerKey` nullish check** (`wallet.ts:997`)

`if (!workerKey)` catches both `undefined` and empty — correct. The error message is clear. ✅

**8. Predicate ordering in `sendMessageWithRetry`** (`wallet.ts:843-850`)

The `isMessageBusInitializingError` check runs **before** `isMessageBusNotInitializedError`, which is critical since the "initializing" message is a superset. This ordering is correct and well-documented in the comment at `wallet.ts:161-166`. ✅

### Test coverage

- **Worker-side:** 8 new tests in `messageBus.test.ts` covering FIFO ordering, init-pending rejection, handler rollback, failed-init recovery, error class hierarchy. Thorough.
- **Page-side:** 5 new tests in `wallet.test.ts` covering identity match, readonly/signing mismatch, missing key, reinitialize mismatch. The `reinitialize` mismatch test correctly verifies `GET_ADDRESS` is never retried against the wrong identity.
- Test harness updated to answer `GET_STATUS` — clean integration with existing tests.

### Cross-repo impact

No public API **removed** or **changed incompatibly**. Two new exports added (`MessageBusInitializingError`, `MESSAGE_BUS_INITIALIZING`) — purely additive. The `MessageBusNotInitializedError` constructor change is backward-compatible via default parameter. No downstream breakage expected.

### Protocol-critical assessment

This PR **does not** touch VTXO handling, transaction signing, forfeit paths, round lifecycle, connector trees, or exit paths directly. It operates at the service-worker message-bus layer. However, the identity-binding race it fixes **is** a loss-of-funds vector (wrong address returned → user deposits to an address they don't control). The fix is correct and well-tested. **Human sign-off recommended** given the security implications, but the code itself is sound.

---

**LGTM.** Solid engineering. The FIFO chain + page-side assertion is the right two-layer defense for this class of race. Ship it.

🤖 Reviewed by Arkana

File: packages/ts-sdk/src/wallet/serviceWorker/wallet.ts [hunk 1]
```diff
@@ -807,10 +833,19 @@ export class ServiceWorkerReadonlyWallet implements IReadonlyWallet {
...
             } catch (error: any) {
+                // An init is already in flight in the worker. Wait for it to
+                // settle and retry the same message; do not enqueue a redundant
+                // re-init behind it. A wait does not consume the re-init budget.
+                if (isMessageBusInitializingError(error)) {
+                    if (initWaits >= MAX_INIT_WAITS) throw error;
+                    await sleep(initWaitBackoffMs(initWaits++));
+                    attempt--;
+                    continue;
+                }
```
[the remark this answers is not in the document]
Comment (pietro909) [reply packages/ts-sdk/src/wallet/serviceWorker/wallet.ts]: It would regress legitimate waits.
At that location the worker is by definition still initializing.

**Review by ghost — COMMENTED** [review]
## Incremental Review — commit c97a4b67

**Reviewer: Arkana (automated, aggressive)**
**Scope: delta since last review (38dbb7d0 → c97a4b67)**

This commit does three things: trims verbose doc comments, merges `sendMessageWithEvents` into `sendMessageWithRetry` (addressing my previous finding #1), and removes several tests deemed "too specific."

---

### ✅ Good changes

**1. `sendMessageWithRetry` / `sendMessageWithEvents` dedup** — `wallet.ts:812-855`

The duplicated retry logic is now unified via an optional `withEvents` parameter. This directly addresses my prior review comment about the maintenance hazard. Clean refactor, callers updated correctly in `settle()`, `restoreWallet()`, `recoverVtxos()`, `renewVtxos()`, `migrateDeprecatedSignerVtxos()`. The `protected` visibility is correct since `ServiceWorkerWallet` (subclass) calls it.

**2. Doc trimming** — All files

Verbose implementation rationale removed from inline comments and JSDoc. The code is self-documenting enough that this is fine. Key behavioral comments retained (e.g., "Never tick while an init is queued or running").

---

### ⚠️ Concern: removed test coverage on security-critical paths

**3. Missing test: `!workerKey` branch in `assertWorkerIdentityMatches`** — `wallet.ts:958-960`

The "create() rejects when the worker cannot prove any identity" test was removed. This was the **only** test covering the `if (!workerKey)` guard in `assertWorkerIdentityMatches()`. That guard prevents the wallet from trusting a worker that reports no identity — a loss-of-funds path. Without a test, a future refactor could accidentally remove or weaken this check with no CI signal.

**Recommendation:** Restore a test for the `workerKey === undefined/null` case, or fold it into the existing "rejects when the worker reports a different identity" test as a second assertion.

**4. Missing test: reinitialize identity mismatch** — previously at end of `wallet.test.ts`

The "reinitialize() rejects on mismatch before retrying the original wallet message" test was removed. This was the only end-to-end test proving that when `reinitialize()` rebinds the worker to a **different** identity mid-session, the original message (e.g., `GET_ADDRESS`) is **not** retried against the wrong identity. This is a critical property — it's the difference between "identity mismatch throws" and "identity mismatch throws **before** serving a wrong address."

**Recommendation:** This test was arguably the most important one in the suite. Consider keeping it or replacing it with something equivalent that asserts the retry-after-reinitialize path respects the identity boundary.

---

### Verdict

The code changes are clean and correct. The test removals weaken coverage on two security-relevant branches. Neither is a ship-blocker on its own (the code is correct today), but both reduce the safety net against future regressions on a loss-of-funds path.

**Not changing my approval** — the core logic is unchanged and sound. But I'd strongly prefer restoring coverage for items 3 and 4 before merge.

🤖 Reviewed by Arkana

**Review by ghost — APPROVED** [review]
## Incremental Review — commit 48f7d180

**Reviewer: Arkana (automated, aggressive)**
**Scope: delta since last review (c97a4b67 → 48f7d180)**

### ✅ Concern addressed

**Reinitialize identity mismatch test restored** — `wallet.test.ts:1586-1637`

This was my top concern from the previous review (item #4). The test correctly:
1. Creates a wallet with matching identity (KEY_A)
2. Forces `getAddress()` down the `MESSAGE_BUS_NOT_INITIALIZED` → reinitialize path
3. Makes the 2nd init report a **different** identity (KEY_B)
4. Asserts the mismatch is caught before retry
5. Asserts `GET_ADDRESS` was only sent once — proving the original message was **not** retried against the wrong identity

This is well-constructed and covers the critical property: identity boundary is enforced on the reinitialize path, not just on initial create.

### Remaining minor gap

The `!workerKey` branch in `assertWorkerIdentityMatches` (`wallet.ts:958-960`) still has no dedicated test (item #3 from my previous review). This is a simple null guard and the risk is low — not blocking.

### Verdict

All substantive concerns addressed. Approving.

⚠️ **Protocol-critical reminder**: This PR touches identity binding for service workers — a loss-of-funds path. Human review and approval required before merge.

🤖 Reviewed by Arkana

**Review by ghost — COMMENTED** [review]
## Incremental Review — commit 8c5ca754

**Reviewer: Arkana (automated, aggressive)**
**Scope: delta since last review (48f7d180 → 8c5ca754)**

Rebased commit. Two changes:

1. **`.toReadonly()` dropped** — `wallet.test.ts:1554`: `SingleKey.fromHex(KEY_A)` now used directly instead of `.toReadonly()`. Correct — `create()` accepts `ReadonlyIdentity | Identity`, and `Identity` satisfies the constraint. Matches real-world usage more closely.

2. **Reinitialize identity mismatch test intact** — `wallet.test.ts:1582-1637`: Same test I reviewed in 48f7d180, unchanged. Still correctly asserts the identity boundary on the reinitialize path.

No new concerns. Previous approval unchanged.

⚠️ **Protocol-critical reminder**: Human review still required before merge — this PR touches identity binding on a loss-of-funds path.

🤖 Reviewed by Arkana

**Review by ghost — APPROVED** [review]
## Incremental Review — commit 6182098c

**Reviewer: Arkana (automated, aggressive)**
**Scope: delta since last review (8c5ca754 → 6182098c)**

### ✅ Clean fix

**`normalizedErrorText()` + space-stripped needles** — `electrum.ts:1028-1034, 1045-1067`

`ws-electrumx-client`'s frame parser joins tokens without spaces, so server errors like `"No confirmed transaction matching…"` arrive as `"Noconfirmedtransactionmatchingtherequestedhashwasfound"`. The old `.includes("no confirmed transaction")` never matched.

The fix is correct:
- `normalizedErrorText()` strips all whitespace and lowercases — now matches regardless of whether the transport mangled spaces.
- Needles updated to space-free equivalents: `"notyetinablock"`, `"notinablock"`, `"notinblock"`, `"noconfirmedtransaction"` — all match their original variants post-normalization.
- `isMissingHeightError` also switched to `normalizedErrorText()` for consistency — no-op since `"missingheight"` was already a single token, but keeps the codebase uniform.
- False-positive risk is negligible: these needles are specific enough that no unrelated error would contain them after whitespace stripping.

**Test** — `electrum.test.ts:417-431`

New test uses the actual mangled string (`"Noconfirmedtransactionmatchingtherequestedhashwasfound"`) observed in CI. Directly reproduces the flake. Good.

### No concerns

This is a pure error-classification fix in the electrum provider — no protocol-critical paths (signing, VTXOs, forfeit, exit) are affected. Both callers (`getTxStatus` line 301, `getBlockTime` line 878) use these guards to treat classification hits as "not yet confirmed," which is the safe direction.

Previous approval stands.

⚠️ **Protocol-critical reminder (unchanged)**: The earlier commits in this PR touch identity binding on a loss-of-funds path. Human review still required before merge.

🤖 Reviewed by Arkana

**Review by ghost — APPROVED** [review]
## Incremental Review — commit 6182098c

**Reviewer: Arkana (automated, aggressive)**
**Scope: delta since last review (8c5ca754 → 6182098c)**

### ✅ Clean fix

**`normalizedErrorText()` + space-stripped needles** — `electrum.ts:1028-1034, 1045-1067`

`ws-electrumx-client`'s frame parser joins tokens without spaces, so server errors like `"No confirmed transaction matching…"` arrive as `"Noconfirmedtransactionmatchingtherequestedhashwasfound"`. The old `.includes("no confirmed transaction")` never matched.

The fix is correct:
- `normalizedErrorText()` strips all whitespace and lowercases — now matches regardless of whether the transport mangled spaces.
- Needles updated to space-free equivalents: `"notyetinablock"`, `"notinablock"`, `"notinblock"`, `"noconfirmedtransaction"` — all match their original variants post-normalization.
- `isMissingHeightError` also switched to `normalizedErrorText()` for consistency — no-op since `"missingheight"` was already a single token, but keeps the codebase uniform.
- False-positive risk is negligible: these needles are specific enough that no unrelated error would contain them after whitespace stripping.

**Test** — `electrum.test.ts:417-431`

New test uses the actual mangled string (`"Noconfirmedtransactionmatchingtherequestedhashwasfound"`) observed in CI. Directly reproduces the flake. Good.

### No concerns

This is a pure error-classification fix in the electrum provider — no protocol-critical paths (signing, VTXOs, forfeit, exit) are affected. Both callers (`getTxStatus` line 301, `getBlockTime` line 878) use these guards to treat classification hits as "not yet confirmed," which is the safe direction.

Previous approval stands.

⚠️ **Protocol-critical reminder (unchanged)**: The earlier commits in this PR touch identity binding on a loss-of-funds path. Human review still required before merge.

🤖 Reviewed by Arkana

**Review by ghost — APPROVED** [review]
## Arkana Code Review — ts-sdk#571

**Verdict: APPROVE** with protocol-critical flag for mandatory human sign-off.

---

### Security Fix Assessment

This PR closes a real loss-of-funds vector: concurrent `INITIALIZE_MESSAGE_BUS` calls could interleave and bind the bus to a stale identity, causing `getBoardingAddress()` / `getAddress()` to return a previous identity's address. Funds sent there would be unrecoverable. The fix is sound and well-structured.

### Changes Reviewed

**1. FIFO init serialization (`messageBus.ts`)**
- `initChain` promise-chains concurrent inits so they execute in enqueue order — no interleaving. ✅
- `pendingInitCount` gate (`acceptsWalletMessages()`) refuses wallet messages the instant a new init is accepted, even while queued behind an older one. ✅
- Chain is self-healing: `this.initChain = run.catch(() => undefined)` — a failed init never blocks the next. ✅
- Init failures are delivered to the page as error responses rather than swallowed or thrown as unhandled worker errors. ✅

**2. Handler rollback on partial start failure (`messageBus.ts:doInit`)**
- Tracks `started` handlers; if a later handler's `start()` throws, all previously started ones are stopped. This prevents a half-initialized handler set from serving requests. ✅

**3. Tick suppression during re-init (`messageBus.ts`)**
- `doInit` cancels any scheduled tick before stopping handlers.
- `runTick` returns early when `pendingInitCount > 0`.
- Prevents stale timers from ticking handlers that are stopped or mid-rebuild. ✅

**4. Page-side identity assertion (`wallet.ts:assertWorkerIdentityMatches`)**
- After init/re-init, verifies the worker's reported `xOnlyPublicKey` matches `this.identity.xOnlyPublicKey()` via GET_STATUS.
- Called in both `create()` and `reinitialize()` — covers initial binding and recovery paths. ✅
- Correctly uses `sendMessageDirect` (no retry) — an assertion should not retry against a mismatched worker.
- Fails safe: if GET_STATUS errors or `xOnlyPublicKey` is absent, throws before returning a usable wallet. ✅

**5. Error differentiation (`errors.ts`)**
- `MessageBusInitializingError extends MessageBusNotInitializedError` — backward-compatible via inheritance and the superset message string. ✅
- Page-side `sendMessageWithRetry` uses this to distinguish "wait for in-flight init" (backoff) from "worker needs re-init" (reinitialize). Prevents the page from forcing a redundant re-init that could worsen the race. ✅
- Backoff is bounded: 8 attempts, capped at 2s per wait (~8.1s total). A stuck init surfaces an error rather than hanging. ✅

**6. `sendMessageWithRetry` merge (`wallet.ts`)**
- `sendMessageWithEvents` folded into `sendMessageWithRetry` via optional `withEvents` parameter. Same behavior for both streaming and direct paths. Timeout semantics preserved: streaming still bypasses `timeoutMs`, direct still applies it. ✅
- `attempt--` on initializing-error ensures wait loops don't consume reinit retries. Correct but slightly non-obvious — the comment is adequate.
- Visibility change `private → protected` is necessary for subclass access. ✅

**7. Electrum space-stripping (`electrum.ts`)**
- `normalizedErrorText()` strips all whitespace before substring matching. Fixes the real CI flake where `ws-electrumx-client`'s frame parser drops spaces from server error messages.
- Needles updated to whitespace-free form. Verified no false-positive substring collisions (e.g., `"notinblock"` does NOT appear as a substring in `"notyetinablock"`). ✅
- `isMissingHeightError` also switched to `normalizedErrorText` for consistency — functionally identical since `"missingheight"` is already space-free. ✅

**8. Test coverage**
- `messageBus.test.ts`: 9 new tests covering FIFO ordering, initializing-error gating, handler rollback, failed-init recovery, stale-tick suppression, and error classification. ✅
- `wallet.test.ts`: Identity boundary assertions for create/reinitialize, mismatch rejection, harness updated with GET_STATUS auto-answer. ✅
- `electrum.test.ts`: Space-stripped error matching. ✅

### No Issues Found

The implementation is clean, well-commented, and covers the edge cases. Cross-repo impact is limited to the new public exports (`MESSAGE_BUS_INITIALIZING`, `MessageBusInitializingError`) which are additive only.

---

⚠️ **PROTOCOL-CRITICAL**: This PR touches identity binding and service-worker wallet initialization — bugs here directly affect which addresses and signing keys the SDK uses. **Human review and approval required before merge.**

🤖 *Reviewed by Arkana*

File: packages/ts-sdk/src/wallet/serviceWorker/wallet.ts
Comment (Dunsin-cyber) [inline packages/ts-sdk/src/wallet/serviceWorker/wallet.ts:222]:
```suggestion
    // RESTORE_WALLET is a streaming/long-running path (sendMessageWithRetry)
```
i changed `sendMessageWithEvents` to `sendMessageWithRetry` since it has been renamed in the file

**Review by Kukks — APPROVED** [review]
